src/backend/distributed/utils/citus_clauses.c

```diff
@@ -0,0 +1,252 @@
+/*
+ * citus_clauses.c
+ *
+ * Routines roughly equivalent to postgres' util/clauses. 
+ *
+ * Copyright (c) 2016-2016, Citus Data, Inc.
+ */
+
+#include "postgres.h"
+
+#include "distributed/citus_clauses.h"
+
+#include "catalog/pg_type.h"
+#include "executor/executor.h"
+#include "nodes/makefuncs.h"
+#include "nodes/nodeFuncs.h"
+#include "nodes/nodes.h"
+#include "nodes/primnodes.h"
+#include "optimizer/planmain.h"
+#include "utils/datum.h"
+#include "utils/lsyscache.h"
+
+static Node * PartiallyEvaluateExpressionWalker(Node *expression, bool *containsVar);
+static Expr * citus_evaluate_expr(Expr *expr, Oid result_type, int32 result_typmod,
+								  Oid result_collation);
+
+
+/*
+ * Walks the expression, evaluating any STABLE or IMMUTABLE functions so long as they
+ * don't reference Vars.
+ */
+Node *
+PartiallyEvaluateExpression(Node *expression)
+{
+	bool unused;
+	return PartiallyEvaluateExpressionWalker(expression, &unused);
+}
+
+
+/*
+ * When you find a function call evaluate it, the planner made sure there were no Vars
+ *
+ * Tell the parent whether you are a Var. If your child was a var tell your parent
+ *
+ * A little inefficient. It goes to the bottom of the tree then calls EvaluateExpression
+ * on each function on the way back up. Say we had an expression with no Vars, we could
+ * only call EvaluateExpression on the top-most level and get the same result.
+ */
+static Node *
+PartiallyEvaluateExpressionWalker(Node *expression, bool *containsVar)
+{
+	bool childContainsVar = false;
+	Node *copy = NULL;
+
+	if (expression == NULL)
+	{
+		return expression;
+	}
+
+	/* pass any argument lists back to the mutator to copy and recurse for us */
+	if (IsA(expression, List))
+	{
+		return expression_tree_mutator(expression,
+									   PartiallyEvaluateExpressionWalker,
+									   containsVar);
+	}
+
+	if (IsA(expression, Var))
+	{
+		*containsVar = true;
+
+		/* makes a copy for us */
+		return expression_tree_mutator(expression,
+									   PartiallyEvaluateExpressionWalker,
+									   containsVar);
+	}
+
+	copy = expression_tree_mutator(expression,
+								   PartiallyEvaluateExpressionWalker,
+								   &childContainsVar);
+
+	if (childContainsVar)
+	{
+		*containsVar = true;
+	}
+	else
+	{
+		copy = EvaluateExpression(expression);
+	}
+
+	return copy;
+}
+
+
+Node *
```

Comment:
Function needs docs. Some explanation about why it's needed (for the type info), would be helpful for later readers as well.  It'd be good to document why you chose exactly these nodes for evaluation, and why you're e.g. not handling RelabelType, ...